Test  tagging

fdkong opened the pull request:
Clean up the @friedmud's code a little so that we can test it to see if it breaks anything. It is better to start with a working code.

friedmud committed:
Add initial support for creating tags refs #9669

friedmud committed:
Can add tagged vectors refs #9669

friedmud committed:
Start allocating sub_res based on tags refs #9669

friedmud committed:
Move Kernel over to using Tags to be able to contribute to multiple residual vectors refs #9669

friedmud committed:
Differentiate matrix and vector tags refs #9669

fdkong committed:
Clean up Derek's tagging code a little

so that we can test it. It is better to work on something that is working.
Othwerwise we need to a lot of time to fix the broken stuff.

Refs #9669, #9737

fdkong committed:
Tagging: GCC+linux has an issue on addVectorTag

tag_id = _vector_tag_name_to_tag_id[tag_name] =   _vector_tag_name_to_tag_id.size();

 _vector_tag_name_to_tag_id[tag_name] is excuted first, and so we end up with tag_id = 1
at the very begining.

Refs #9669, #9737

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/125863/) on d6601ba wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10111/site)

This comment will be updated on new commits.

fdkong commented:
@YaqiWang, do you have any idea why Rattlesnake fails?

YaqiWang commented:
I probably know, but need to double check. That test is using a fake transient executioner.

YaqiWang commented:
That executioner does not setup a time integrator, which will call `residualVector` to add a vector for time kernels to tag. The fix should be trivial. We can work it out tomorrow.

YaqiWang commented on framework/src/base/SubProblem.C:
```diff
@@ -46,6 +46,60 @@ SubProblem::SubProblem(const InputParameters & parameters)
+TagID
+SubProblem::addVectorTag(TagName tag_name)
+{
+  if (_vector_tag_name_to_tag_id.find(tag_name) == _vector_tag_name_to_tag_id.end())
```
It is better to have something like
```
auto it = _vector_tag_name_to_tag_id.find(tag_name);
if (it == _vector_tag_name_to_tag_id.end())
{
   ...
   return tag_id;
}
else
  return it->second;
```
This piece of code is not efficiency critical but set a bad example of using map.

YaqiWang commented on test/tests/kernels/tag_errors/tag_doesnt_exist/bad_transient.i:
```diff
@@ -0,0 +1,50 @@
+
+[Executioner]
+  # Preconditioned JFNK (default)
+  type = Steady
```
so `Steady::checkIntegrity()` will not be never reached?

fdkong replied:
Right. I think so.

YaqiWang commented on framework/include/utils/MooseTypes.h:
```diff
@@ -161,8 +163,8 @@ enum VarKindType
 {
-  KT_TIME = 0,
-  KT_NONTIME = 1,
+  KT_NONTIME = 0,
```
just curious, why is this change?

fdkong replied:
We want tag id for KT_NONTIME be zero because it is default tag.

YaqiWang commented on framework/src/base/NonlinearSystemBase.C:
```diff
@@ -145,6 +143,13 @@ NonlinearSystemBase::NonlinearSystemBase(FEProblemBase & fe_problem,
+  _Re_non_time = &addVector(_Re_non_time_tag, false, GHOSTED);
+
+  // Don't need to add the matrix - it already exists (for now)
+  _Ke_non_time_tag = _fe_problem.addMatrixTag("NONTIME");
```
Maybe assign this tag to the system matrix in `_tagged_matrices `?

fdkong replied:
good point

YaqiWang commented on framework/src/kernels/Kernel.C:
```diff
@@ -41,21 +41,62 @@ Kernel::Kernel(const InputParameters & parameters)
     _u_dot(_var.uDot()),
     _du_dot_du(_var.duDotDu())
 {
+  auto & vector_tag_names = getParam<MultiMooseEnum>("vector_tags");
```
We will possibly need the same thing for BCs, DGKernels, etc.

fdkong replied:
We want a tagging interface, and then BCs, DGKernel, etc. should derive from the tagging interface.

YaqiWang commented on framework/src/base/NonlinearSystemBase.C:
```diff
@@ -145,6 +143,13 @@ NonlinearSystemBase::NonlinearSystemBase(FEProblemBase & fe_problem,
+  // Don't need to add the matrix - it already exists (for now)
+  _Ke_non_time_tag = _fe_problem.addMatrixTag("NONTIME");
+
+  _u_dot = &addVector("u_dot", true, GHOSTED);
```
I think it is time to add this vector only for transient. MooseVariable can adapt easily.

YaqiWang replied:
We could add the vector in `solutionUDot()` like `_Re_time`.

fdkong replied:
The solutionUDot may never be tagged.

fdkong committed:
SubProblem, Tagging: More efficient to use std::map

Refs #9669

fdkong committed:
Tagging: add capability to remove tagged vectors and matrices

Refs #9669

fdkong committed:
EigenProblem: add a callback compute Ax and Bx together

Refs #7398 #9669

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/129124/) on 0bab2f5 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 6067c3c36590fbe2bce980fe989256732cb38e79`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10111/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`

fdkong commented:
@YaqiWang, could you fix Rattlesnake on this? I am already used to ignore Rattlesnake because of this error.  I am gradually adding new features. It is a good idea to make sure everything work during this process.

YaqiWang commented:
Try adding this
```
  _nl.residualVector(Moose::KT_TIME);
```
in `RecoverExodus.C`. Let me stop by and working with you on this.

fdkong committed:
Tagging: Nodal BC supports tags

Refs #9669

fdkong committed:
Tagging: Assembly all vectors together

This is already working for eigenvalue problems right now

Refs  #7398 #9669

moosebuild commented:
Job [TBB Test PETSc master](https://www.moosebuild.org/job/131013/) on 9b68fc9 : invalidated by @fdkong

fdkong commented:
OK..., it is time to look at the design now. I have been working on the tagging capability a little bit. And it is actually already working for eigenvalue problems right now. For generalized eigenvalue problems Ax = \lambda Bx, we can form Ax and Bx simultaneously right now, and it is supposed to be more efficient. 

I think there are several things we would like to discuss:

(1) How to manage tagged  matrices and tagged vectors? Now, those tagged matrices and vectors are stored in `system\libmesh`, and their corresponding pointers are managed in `systembase` and `subproblem`. I honestly do not know if it is a right way or not.  @aeslaughter may have a better idea.

(2)` Tagging interface`. We may not like the current implementation with which for each kernel-like class, we handle input parameters and associate tags with that kernel.  A lot of things (not only kernel) contribute to the residual or the Jacobian, and all of them should be tagged. We therefore may have a generic tagging class, and all kernel-like classes such as kernel, nodal-kernel, nodalBC, DGKernel and so on should derive from the generic  tagging class. So we can handle all tagging things in one single place. 
```

                        |   Kernel 
tagging interface  <---   NodalKernel 
                        |   DGKernel
                        |    ......
```

(3) Right now, users set tags through input files. We may also want to add some public APIs to support advanced users, and they can set tags in `action` or something like that.


Ping @idaholab/moose-team

YaqiWang commented on framework/src/base/EigenProblem.C:
```diff
@@ -190,3 +190,15 @@ EigenProblem::computeResidualType(const NumericVector<Number> & soln,
+                                    NumericVector<Number> & Bx,
+                                    Moose::KernelType type)
+{
+  FEProblemBase::computeResidualType(soln, Ax, type);
```
now passing `Ax` in this function does not make sense because of tagging. Instead, we should have a new function to associate a vector to a tag like `setTaggedVector(TagName tag, NumericVector<Number> & x)`. I am also curious who should be creating the vector, PETSc or MOOSE?

fdkong replied:
Good question. If I let PETSc create vectors, I would worry about MOOSE manipulate the vectors in arbitrary way. This potentially be very buggy.  I

I would like to have MOOSE create their own vectors, and PETSc also has their own vectors. Once the MOOSE vectors are filled, they will be copied over to the PETSc vectors. In this way, I won't worry about MOOSE changes things inside of PETSc. The disadvantage is that we hold extra copies. The number of tagged vectors are supposed to be small.

fdkong replied:
Any suggestions are welcome on this.

YaqiWang commented on framework/src/base/NonlinearEigenSystem.C:
```diff
@@ -220,6 +226,13 @@ NonlinearEigenSystem::getNthConvergedEigenvalue(dof_id_type n)
+void
+NonlinearEigenSystem::computeResidualClose(NumericVector<Number> & Ax, NumericVector<Number> & Bx)
+{
+  Ax = getVector(this->_Ax_tag);
```
If we have `setTaggedVector`, we can avoid this copy operation.

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -526,6 +526,32 @@ mooseSlepcEigenFormFunctionB(SNES snes, Vec x, Vec r, void * ctx)
+  // update local solution
+  X_global.localize(*sys.current_local_solution.get());
+
+  AX.zero();
```
These `zero` and `close` operations can be moved into `computeResidualABType`.

fdkong replied:
I like to keep this way because it consistent with elsewhere.

YaqiWang commented on framework/src/bcs/NodalBC.C:
```diff
@@ -35,6 +36,18 @@ validParams<NodalBC>()
       "contributions to.  Everything about that variable must match everything "
       "about this variable (the type, what blocks it's on, etc.)");
 
+  // These are the default names for tags, but users will be able to add their own
```
yes, I agree these should be provided by a `TagInterface`.

YaqiWang replied:
I also think for now, we possibly should make these two parameters private and only MOOSE `System` can set them.

fdkong replied:
I think we need some actions to handle these parameters.

YaqiWang replied:
This is a question we should discuss with other moose members like @permcody

YaqiWang commented:
I do not want to give up my right to review this and I have done. If somebody wants, he should let the team know so that we do not have to wait. Thanks.

fdkong commented on test/tests/problems/eigen_problem/ne.i:
```diff
@@ -24,12 +24,14 @@
     type = Reaction
     variable = u
     eigen_kernel = true
+    vector_tags = 'nontime Bx_tag'
```
Right,

fdkong replied:
Action may be even better

fdkong commented on test/tests/problems/eigen_problem/ne.i:
```diff
@@ -39,6 +41,7 @@
     variable = u
     boundary = '0 1 2 3'
     value = 0
+    vector_tags = 'nontime Ax_tag Bx_tag'
```
Bx also has boundary conditions.

YaqiWang replied:
So this is true only for nodal boundary conditions, right?

permcody commented:
Tag @friedmud - You will probably want to look at this after Fande has worked on it. I have looked at it yet.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/132342/) on 228a98f wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 5c302b9849d2cc6c32d08afc33fef9d5b3b2bf11`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10111/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`

fdkong committed:
Add TaggingInterface.{C,h}

to handle tagging related stuffs. Other kernel-like classes
should derive from this.

Refs  #7398 #9669

fdkong committed:
Add TaggingInterface to BoundaryCondition

Refs  #7398 #9669

fdkong committed:
Add new APIs in TaggingInterface.{C,h}

This allows us to assign tags to kernel or BC from NonlinearSystem,
NonlinearEigenSystem or Executioner.

Refs  #7398 #9669

fdkong committed:
Nodal BC for the right size of eigenvalue problems

Refs #7398

fdkong committed:
Adjust tagging to make all eigenvalue examples work

All eigenvalue examples work with tagging, but require explicitly
add tags.

We will remove this limitation in the future

Refs  #7398 #9669

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/136079/) on 2bb5dad wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 2eed417cb3fdc61d1bd44b4fbc921dba9915dfb0`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10111/style.patch).
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`

fdkong committed:
Eigenvalue solver: clean up the old implementation

Part of the old implementation is cleaned up using the tagging
capability

Refs #7398

fdkong commented:
@permcody, this PR is ready for you guys to review. Tagging capability works for eigenvalue problems, and this is all I want to do at the first stage. It might be enhanced for explicit method and other stuffs, in the future, but I am not going to do that right now.

My first priory is to migrate Rattlesnake eigenvalue solvers for @YaqiWang.

permcody commented:
Hmm, not quite - this breaks Tiamat. I think all the other failures might be ok.

fdkong commented:
I think  it is ready, now. 

Tiamat should fix the restart logic instead of using the hard-coded vector names.

fdkong committed:
Subproblem: Changed {getMatrixTag,getVectorTagID}

to {getMatrixTagID, getVectorTagID}

Refs  #7398 #9669

friedmud commented:
Need to create a slightly different interface for `FEProblem::computeResidualType()`... it should take a vector to fill up and a set of tags... compute the residual into all of the tagged vectors then sum them into the ultimate vector.

There should also be a function that just computes into the tagged vectors without computing the ultimate summed residual.

permcody commented:
Extending the 30-day limit since @fdkong is out of the country. We want this capability.

fdkong commented:
Will be on this soon!

moosebuild commented:
Job [Test Tiamat](https://www.moosebuild.org/job/152316/) on 406cd5e : invalidated by @permcody

Test has been skipped

fdkong committed:
KernelWarehouse: added taggedkernelWarehouse

Refs #10366

moosebuild commented:
Job [Test debug SLEPc](https://www.moosebuild.org/job/153486/) on 21d1755 : invalidated by @fdkong

fdkong committed:
Remove Moose::KT_TIME from TimeKernel

Refs #10366

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/154016/) on a9ea823 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7355abbe7cd56969185b85ab49b2a01c9926b8ee`

fdkong committed:
Remove Moose::KT_TIME from ScalarKernel and its derivatives

Refs #10366

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/154350/) on 25b5d31 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7355abbe7cd56969185b85ab49b2a01c9926b8ee`

fdkong committed:
Remove Moose::KT_TIME from NodalKernel

Refs #10366

fdkong committed:
Remove Moose::KernelType from Assembly{C,h}

Refs #10366

moosebuild commented:
Job [Pthreads Test](https://www.moosebuild.org/job/154844/) on 48a0750 : invalidated by @brianmoose

updated recipe

fdkong committed:
Remove Moose::{KT_TIME, KT_NONTIME} from DisplacedProblem

and FEProblemBase

Refs  #10366

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/154927/) on ac02711 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7355abbe7cd56969185b85ab49b2a01c9926b8ee`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/154945/) on 389fd58 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7355abbe7cd56969185b85ab49b2a01c9926b8ee`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/154975/) on 4f11b0a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7355abbe7cd56969185b85ab49b2a01c9926b8ee`

moosebuild commented:
Job [Modules debug](https://www.moosebuild.org/job/155006/) on 9ad9a9f : invalidated by @fdkong

Timeout

moosebuild commented:
Job [Test Intel](https://www.moosebuild.org/job/155016/) on 9ad9a9f : invalidated by @fdkong

recover_3

fdkong committed:
Replace kerneltype with  warehouse

Refs #10366

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/156612/) on 4a226ba wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 77b4d5e33a8fe669bc0f7203e45f2ef661f39687`

fdkong committed:
Delete computeJacobian from EigenProblem

because the extra resiudal computation has been removed by
PR #10597

fdkong commented:
@YaqiWang, could you take a look at Rattlesnake?

YaqiWang commented:
Feel free to stop by. We can communicate what we are doing.

fdkong commented:
Ready for review @friedmud

friedmud commented on framework/src/kernels/KernelWarehouse.C:
```diff
@@ -15,7 +15,14 @@
-KernelWarehouse::KernelWarehouse() : MooseObjectWarehouse<KernelBase>() {}
+KernelWarehouse::KernelWarehouse()
+  : MooseObjectWarehouse<KernelBase>(),
+    _vector_tag_to_kernel_warehourse(_num_threads),
```
`warehouse`

friedmud commented on framework/include/kernels/KernelWarehouse.h:
```diff
@@ -49,9 +49,29 @@ class KernelWarehouse : public MooseObjectWarehouse<KernelBase>
    */
   virtual void updateActive(THREAD_ID tid = 0) override;
 
+  KernelWarehouse & getVectorTagKernelWarehouse(TagID tag_id, THREAD_ID tid);
```
Make `KernelWarehouse` inherit from `MooseObjectTagWarehouse`

friedmud commented on framework/src/base/Assembly.C:
```diff
@@ -1396,42 +1431,45 @@ Assembly::cacheResidualNodes(DenseVector<Number> & res, std::vector<dof_id_type>
   {
-    _cached_residual_values[type].push_back(res(i));
-    _cached_residual_rows[type].push_back(dof_index[i]);
+    _cached_residual_values[0].push_back(res(i));
```
We can't have `0` here like this.  If this should be `nontime` then we need to pull out the `nontime` tag.  Even better: we should pass the tag into this function.

friedmud commented on framework/src/base/Assembly.C:
```diff
@@ -1396,42 +1431,45 @@ Assembly::cacheResidualNodes(DenseVector<Number> & res, std::vector<dof_id_type>
+  mooseAssert(tag_name_to_tag_id.size() == _sub_Re.size(),
+              "the number of tags does not equal to the number of residuals ");
+
+  for (auto tag_it = tag_name_to_tag_id.begin(); tag_it != tag_name_to_tag_id.end(); ++tag_it)
```
Use C++11 for each

friedmud commented on framework/src/base/ComputeJacobianThread.C:
```diff
@@ -179,6 +152,18 @@ ComputeJacobianThread::subdomainChanged()
+
+  // If users pass a empty vector or a full size of vector,
+  // we take all kernels
+  if (!_tags.size() || _tags.size() == _fe_problem.numVectorTags())
```
This whole `if` statement is a good idea!  Way to think about optimizing the usual case (all or 1) 👍

fdkong committed:
Deprecate APIs computeResidualType() and residualVector()

Refs #9696

friedmud commented on framework/src/base/FEProblemBase.C:
```diff
@@ -998,19 +995,14 @@ FEProblemBase::prepareAssembly(THREAD_ID tid)
+  auto & tags = getVectorTag();
+
+  for (auto & tag : tags)
+    if (_nl->hasVector(tag.second))
```
Why would you ever have the tag - but not have the vector?

friedmud commented on framework/src/base/FEProblemBase.C:
```diff
@@ -1019,21 +1011,24 @@ FEProblemBase::addResidual(THREAD_ID tid)
-    _assembly[tid]->addResidualNeighbor(residualVector(Moose::KT_NONTIME), Moose::KT_NONTIME);
+  auto & tags = getVectorTag();
+
+  for (auto & tag : tags)
```
Can this tag loop be done inside assembly?

friedmud commented on framework/src/base/TaggingInterface.C:
```diff
@@ -0,0 +1,176 @@
+}
+
+void
+TaggingInterface::addVectorTag(TagName & tag_name)
```
`addVectorTag` is not good to use here... we use "add" to me "create something new".  From the name I would assume that this function will actually add a new tag to the system.

How about `usesVectorTag` or something like that?

friedmud commented on framework/src/base/TaggingInterface.C:
```diff
@@ -0,0 +1,176 @@
+}
+
+void
+TaggingInterface::appendTaggedLocalResidual()
```
`append` isn't really the right word for this.  How about `accumulate`?

friedmud reviewed (changes requested):
Looking good - just a few changes so far (might find more later)

fdkong committed:
Docs for tagging

Refs #9669

moosebuild commented:
Job [App tests](https://www.moosebuild.org/job/159103/) on af78e8f : invalidated by @fdkong

moosebuild commented:
Job [Unity build](https://www.moosebuild.org/job/159105/) on af78e8f : invalidated by @fdkong

moosebuild commented:
Job [App tests](https://www.moosebuild.org/job/159103/) on af78e8f : invalidated by @fdkong

friedmud committed:
Remove KernelWareHouse refs #10366

fdkong committed:
This may work for multiple matrices and vectors

This works for most of existing tests

Will create new tests to make sure (BCs and Kernels) work for
 multiple matrices and vectors

Refs #9669

fdkong committed:
Seems working with multiple vectors

Refs #9669

fdkong committed:
Seems working with multiple matrices

Refs #9669

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/163216/) on db4007d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0cb09bf3c92d5b65b9ea654c215e5f24be5ac053`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/163233/) on 9a654ba wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0cb09bf3c92d5b65b9ea654c215e5f24be5ac053`

fdkong committed:
A tag vector test

BCs and Kernels are associated with several tags

Refs #9669

fdkong committed:
A tag-matrix test

BCs and Kernels are associated with several matrix tags

Refs #9669

fdkong committed:
Fixed eigensolver for new-tag capability

Refs #7398  #9669

fdkong committed:
We assemble A and B together for linear generalized

eigenvalue problem.

Refs #7398  #9669

fdkong committed:
Add missing the input file and mesh file for

the tag tests

fdkong committed:
Move view tag to outputs

This function is very useful when developing the code

Refs #9669

fdkong committed:
Added tag tests for multiple vectors and matrices

Refs #9669

fdkong committed:
Add registerMooseObject in TagMatrixAux

and TagMatrixAux

Refs #9669

fdkong committed:
Fixed nonunity compiling errors

for CreateProblemAction and ViewTagVector

Refs #9669

fdkong committed:
Rename ComputeResidual{s} and ComputeJacobian{s}

to avoid [-Werror=overloaded-virtual]

Refs #9669

fdkong committed:
Mat.zero_rows() should be called by everyone

Refs #9669

fdkong committed:
MooseObjectTagWarehouse: add Threads::spin_mutex for

std::map because it is not thread-safe

Refs #9669

fdkong committed:
TagVectorAux and TagMatrixAux work for

threading.

Refs #9669

fdkong committed:
Fixed threading issues in the warehouse

We made each thread have its own data. It is not a good fix.
But we can not figure out other ways because warehouse is already
messed up.

Threading race.

Refs #9669

fdkong committed:
Register TagTestProblem using the new way

Refs #9669

fdkong committed:
Update documentation for the tagging system

Refs #9669

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/168457/) on a100ebc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bf1bd112e73f907a3bb285d2e5c64c5b8d3b15a4`

fdkong committed:
Fixed test: nodal_kernel_with_aux_var

TimeDerivativeNodalKernel has to work with Transient Executioner

Refs #9669

moosebuild commented:
Job [Modules debug PETSc master](https://civet.inl.gov/job/168568/) on 5460db1 : invalidated by @fdkong

SLEPc fixed

moosebuild commented:
Job [Test PETSc master](https://civet.inl.gov/job/168570/) on 5460db1 : invalidated by @fdkong

SLEPC fixed

moosebuild commented:
Job [Test debug PETSc master](https://civet.inl.gov/job/168569/) on 5460db1 : invalidated by @fdkong

SLEPc fixed

fdkong committed:
Make some tests work with petsc release

These tests are not designed well, and they are very sensitive to petsc changes

It does not make sense to test against petsc-master because petsc-master changes
everyday

DIFF materials/material_dependency.dont_reinit_mat_for_aux FAILED (EXODIFF)
DIFF multiapps/picard_catch_up.test FAILED (EXODIFF) [max_cpus=1]
DIFF multiapps/picard_failure.test FAILED (EXODIFF) [max_cpus=1]
DIFF outputs/iterative.csv FAILED (CSVDIFF) [max_cpus=1]

Refs #9669

fdkong committed:
Fixed compiler errors because of rebase

Refs #9669

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/171465/) on 9112281 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 130e3ad74aef3bf89b6d47f7e1d2c8a3b60b2c00`

friedmud committed:
Const correctness refs #10837

friedmud committed:
More const correctness and fixup a few API functions refs #10837

fdkong committed:
After rebase, make VectorNodalBC work for tagging system

Refs #9669

fdkong committed:
Restore computeResidualType that will

take the boundary (residual tag) into consideration as well

Refs #9669

fdkong committed:
Make BC computeResidual(NumericVector<Number> & /*residual*/) backward compatible

Some animals code for example bighorm acutally override this function

Refs #9669

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/175224/) on 5dce7e5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 424301c110c761e239a28c041c3ed3723d6c55ce`

fdkong committed:
Added using NodalBCBase::computeResidual to avoid compiler warning

Refs #9669

friedmud committed:
Edit and update documentation refs #10111

friedmud committed:
Move new files to fit the new directory structure refs #10111

fdkong committed:
Fixed compiler errors after rebase

Refs #9669

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/176031/) on 29bf6ee wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7d4fe54e034bcc2723b7d3f05c43eaf1d41d8ae5`

fdkong commented:
@friedmud,  I think it is ready right now.

fdkong committed:
Don't like ViewTagVector.{C,h} anymore

Refs #9669

fdkong committed:
Clean up some APIs for tagging

Refs #9669

fdkong committed:
Added "missing" disassociateXXXFromTag

Refs #9669

rwcarlsen commented on framework/include/problems/SubProblem.h:
```diff
@@ -68,6 +68,80 @@ class SubProblem : public Problem
+   */
+  virtual TagName vectorTagName(TagID tag);
+
+  virtual std::map<TagName, TagID> & getVectorTag() { return _vector_tag_name_to_tag_id; }
```
Should this function name be plural since it can return info for several tags?  It confused me at first at call locations.

rwcarlsen commented on test/tests/misc/check_error/old_integrity_check.i:
```diff
@@ -31,10 +31,11 @@
-    type = TimeDerivative
-    variable = v
-  [../]
+#  This will generate it's own error now
```
Just delete the test rather than comment out.

rwcarlsen commented on test/tests/misc/check_error/tests:
```diff
@@ -261,11 +261,13 @@
+#    input = 'invalid_steady_exec_test.i'
+#    expect_err = "You have specified time kernels in your steady state simulation"
+#    deleted = true
+#  [../]
```
remove this?

rwcarlsen commented on test/tests/kernels/tag_errors/tag_doesnt_exist/tests:
```diff
@@ -0,0 +1,21 @@
+  [./bad_transient]
+    type = 'RunException'
+    input = 'bad_transient.i'
+    expect_err = 'was assigned an invalid vector_tag'
```
Let's add requirement, design and issues tags here.

rwcarlsen commented on test/tests/kernels/tag_errors/tag_doesnt_exist/tests:
```diff
@@ -0,0 +1,21 @@
+    input = 'bad_tag.i'
+    cli_args = 'Kernels/diff/vector_tags=nope'
+    expect_err = 'was assigned an invalid vector_tag'
+  [../]
```
Let's add requirement, design and issues tags here.

rwcarlsen commented on test/tests/kernels/tag_errors/tag_doesnt_exist/tests:
```diff
@@ -0,0 +1,21 @@
+    input = 'bad_tag.i'
+    cli_args = 'Kernels/diff/matrix_tags=nope'
+    expect_err = 'was assigned an invalid matrix_tag'
+  [../]
```
Let's add requirement, design and issues tags here.

rwcarlsen reviewed (changes requested):
Well - It's not the worst...  Mostly just more/better doc comments and test spec requirement+design param entries, etc.

friedmud commented on framework/src/problems/DisplacedProblem.C:
```diff
@@ -609,11 +674,14 @@ DisplacedProblem::addCachedResidual(THREAD_ID tid)
 }
 
+// This is a really bad API
+// I hate this API
```
Remove this comment (or change the API)

friedmud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1073,13 +1057,14 @@ FEProblemBase::addCachedResidual(THREAD_ID tid)
     _displaced_problem->addCachedResidual(tid);
 }
 
+// Hate this API
```
Remove this comment (or change the API)

friedmud reviewed (changes requested):
Some comments to work on...

fdkong committed:
Upgrade SLEPc version control in SlepcSupport.C

Refs #9669

fdkong committed:
Added more documentation for tagging system and

improve some APIs

Refs #9669

fdkong commented on framework/include/bcs/NodalBCBase.h:
```diff
@@ -37,7 +37,13 @@ class NodalBCBase : public BoundaryCondition,
-  virtual void computeResidual(NumericVector<Number> & residual) = 0;
+  virtual void computeResidual(NumericVector<Number> & /*residual*/)
+  {
+    ;
```
It is temporarily kept here for bighorn. It should be deleted after this PR is merged and bighorn is fixed.

fdkong committed:
Refresh Eigenvalue solvers

Refs #7398 #9669

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/177636/) on 853f9a9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e24c3244430c115f338e7763146ec541f39fc45e`

fdkong commented:
@friedmud, Do you want to review again ?

fdkong commented:
Eigenvalue solvers are refreshed, and the flags `eigen_kernel` and `eigen_bc` are removed.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/178787/) on 40e0e92 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/10111/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/10111/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5962d4dc56578db13bdd96febf454faf5f272834`

friedmud reviewed (changes requested):
Ok - just a few small things left.

fdkong committed:
Improve TaggingInterface{C,h}

Refs #9669
